ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -482,68 +492,89 @@ KEXT_STATIC int HandleFileOpOperation(
         
         bool isDirectory = (0 != vnode_isdir(currentVnode));
         
-        VirtualizationRootHandle root = RootHandle_None;
+        VirtualizationRootHandle toRoot = RootHandle_None, fromRoot = RootHandle_None;
         FsidInode vnodeFsidInode = {};
-        int pid;
-        if (!ShouldHandleFileOpEvent(
+        int pid = -1;
+        bool messageDestinationProvider = ShouldHandleFileOpEvent(
                 &perfTracer,
                 context,
                 currentVnode,
                 action,
                 isDirectory,
-                &root,
+                fromPath,
+                &toRoot,
+                &fromRoot,
                 &vnodeFsidInode,
-                &pid))
+                &pid);
+        bool messageSourceProvider = VirtualizationRoot_IsValidRootHandle(fromRoot);
+        
+        if (!messageDestinationProvider && !messageSourceProvider)
```

Comment:
To be clear if ShouldHandleFileOpEvent is false, but fromRoot is a valid root handle we continue?
We have a number of checks in ShouldHandleFileOpEvent (ex check the pid to avoid deadlocks).  Are we sure this is a safe change?

Reply:
Hmm, I guess in the case where a provider process is moving (or hardlinking) a file from one virtualisation root to another, I'm not sure we should be messaging the provider for the other virtualisation root. Thoughts? I mean it's not exactly a likely scenario. And to be clear, the code as it stands should never self-message (i.e. sending a message to the provider that caused the event.)

Reply:
Am I correct that the case we're trying to catch is:
A is in the Virtualization Root
B is out of the Virtualization Root

`ln A B` is called.  Now when B is changed A would be updated too, but without any reference to A.  Hence we want to make sure A gets on our watch (modified paths) list.

If so I would think ShouldHandleFileOpEvent should cover it all.  

Or are we talking about a case where
A is in 1 Virtualization Root
B is in a 2nd Virtualization Root

Reply:
Looks like we're in agreement on what *should* happen, and that's certainly what the code so far should be doing. (If you've spotted a discrepancy, please point it out.)

The `ShouldHandle…` functions are suboptimal in a lot of ways to say the least. My original thought was that fixing them in this PR would be too much. I'm sure I can be persuaded otherwise though. :-)

Reply:
Looking into the simple 2-way split suggestion after discussing it out of band, I don't see a super simple way to do this:

 * We don't want to send a message to either the target vnode's provider or the `fromPath`'s provider if the `ShouldIgnoreVnodeType()` or `VirtualizationRoot_VnodeIsOnAllowedFilesystem()` checks fail, and the original version of `ShouldHandleFileOpEvent()` made no distinction between those checks failing versus the vnode not being in a virtualisation root or the PID matching the provider of the root it's in. So we'd either need to do those checks twice or change the API of the function somehow after all, in which case we may as well stick with what we've got and refactor it properly later.

Reply:
@wilbaker and I talked offline about how we can make ShouldHandleFileOpEvent easiest to understand / least complex.  As you've pointed out there is so much we want to refactor about ShouldHandleFileOpEvent.

What we were thinking would be do something like this:
```
   bool shouldHandleTo = ShouldHandleFileOpEvent(other params…, vnode, nullptr, other params..)
   bool shouldHandleFrom = ShouldHandleFileOpEvent(other params…, vnode, frompath, other params..)
```
Where the only difference in `ShouldHandleFileOpEvent` versus what was originally there would be an `if `to choose between the vnodeCache and ActiveProvider_FindForPath for finding the virtualization root.    
This does mean we'd do some duplicate checks, but I'd happily take that if it's easier to follow.

I started that approach [here](https://github.com/Microsoft/VFSForGit/pull/1080/commits/1411818767122fafc46e1b7c90109061d6a49901).  Let me know what you think.  Obviously this commit is a big WIP.


I *think* and correct me if I'm wrong please, what we have now works because we say:
```
        bool messageDestinationProvider = ShouldHandleFileOpEvent( 
                &perfTracer,
		context, 
                currentVnode, 
                action, 
                isDirectory, 
                &root, 
                fromPath, 
                &toRoot, 
                &fromRoot, 
                &vnodeFsidInode, 
                &pid); 
bool messageSourceProvider = VirtualizationRoot_IsValidRootHandle(fromRoot);
```
If we have an invalid vnode, fromRoot would be null, and hence the `VirtualizationRoot_IsValidRootHandle` would return false.  I think that works just trying to think of ways to make the checks here clearer.